### DISCUSSION TITLE
Inquiry About the Use of Shift Operation on `is_eos_codes` in Code

### DISCUSSION THREAD

**Discussion by qixuema** [lucidrains/meshgpt-pytorch#16] [Q&A]
Hello, Phil

I'm currently exploring your code and have a question regarding the implementation of `is_eos_codes`. Specifically, I'm curious about the rationale behind using the `shift` operation on `is_eos_codes` [here](https://github.com/lucidrains/meshgpt-pytorch/blob/231c658bd3a4d38a87fafd4fa3c93bd816ca410e/meshgpt_pytorch/meshgpt_pytorch.py#L1088C20-L1088C20).
```python
shifted_is_eos_tokens = F.pad(is_eos_codes, (1, -1))
mask = shifted_is_eos_tokens.float().cumsum(dim = -1) >= 1
codes = codes.masked_fill(mask, self.pad_id)
```

From my understanding, the purpose of the shift operation is to feed the `eos_token_id` into the `decode_from_codes_to_faces` for processing. However, I am somewhat confused about the potential problems this might introduce. Given that the `eos_token_id` is 16384, and the length of the codebooks is also 16384, this might lead to an indexing issue during the subsequent execution of `autoencoder.decode_from_codes_to_faces`. 
```python
face_coords, face_mask = self.autoencoder.decode_from_codes_to_faces(codes)
```
Specifically, when performing the `gather` operation [here](https://github.com/lucidrains/vector-quantize-pytorch/blob/72d15768bc9d34bbbed36ce5e8dc8deb7222b9bf/vector_quantize_pytorch/residual_vq.py#L107), it might be impossible to find an element at index 16384.

Could you please clarify if my understanding is correct, or if there's something I'm missing here? Any insights or explanations on this would be greatly appreciated.

Thank you for your time and assistance.

Best regards,
Xueqi

**Comment by lucidrains** [comment]
yes you are right, that was my mistake

thank you!
